### PR TITLE
Ports tool_act

### PR DESCRIPTION
**Pull request by Jalleo**
Adds a new defines folder called tools.dm this is for setting behaviours for tools at DEFINE level which is stored as a var on each tool.

ensured the items already ported are converted to use this system as they were having locally set proc/wrench_act etc etc.

Why It's Good For The Game
Less istype() and seperate procs for ALL the tools (eventually)

## Jalleo
:cl:
code: Tools got a slight change only affects vents so far. Please report any issues with welders and plasmacutters on the vents
/:cl:

### CODE REVIEW

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 1]
```diff
@@ -27,6 +27,8 @@
 	matter = list("metal" = 150)
 	origin_tech = "materials=1;engineering=1"
 	attack_verb = list("bashed", "battered", "bludgeoned", "whacked")
+	tool_behaviour = TOOL_WRENCH
+	toolspeed = 1
```
Comment (Rohesie) [inline code/game/objects/items/tools/maintenance_tools.dm]: This is `1` by default already. No need to redefine it.

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 2]
```diff
@@ -46,6 +48,8 @@
 	throw_range = 5
 	matter = list("metal" = 75)
 	attack_verb = list("stabbed")
+	tool_behaviour = TOOL_SCREWDRIVER
+	toolspeed = 1
```
Comment (Rohesie) [inline code/game/objects/items/tools/maintenance_tools.dm]: This is `1` by default already. No need to redefine it.

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 3]
```diff
@@ -107,6 +111,8 @@
 	attack_verb = list("pinched", "nipped")
 	sharp = IS_SHARP_ITEM_SIMPLE
 	edge = 1
+	tool_behaviour = TOOL_WIRECUTTER
+	toolspeed = 1
```
Comment (Rohesie) [inline code/game/objects/items/tools/maintenance_tools.dm]: This is `1` by default already. No need to redefine it.

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 4]
```diff
@@ -141,6 +147,8 @@
 	throw_speed = 1
 	throw_range = 5
 	w_class = 2.0
+	tool_behaviour = TOOL_WELDER
+	toolspeed = 1
```
Comment (Rohesie) [inline code/game/objects/items/tools/maintenance_tools.dm]: This is `1` by default already. No need to redefine it.

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 5]
```diff
@@ -391,6 +399,8 @@
 	origin_tech = "engineering=1"
 	attack_verb = list("attacked", "bashed", "battered", "bludgeoned", "whacked")
 	pry_capable = IS_PRY_CAPABLE_CROWBAR
+	tool_behaviour = TOOL_CROWBAR
+	toolspeed = 1
```
Comment (Rohesie) [inline code/game/objects/items/tools/maintenance_tools.dm]: This is `1` by default already. No need to redefine it.

File: code/game/atoms.dm [hunk 6]
```diff
@@ -605,5 +605,61 @@ Proc for attack log creation, because really why not
...
+			return welder_act(user, I)
+		if(TOOL_WELD_CUTTER)
+			return weld_cut_act(user, I)
+		if(TOOL_ANALYZER)
+			return analyzer_act(user, I)
+
+
+// Tool-specific behavior procs. To be overridden in subtypes.
+/atom/proc/crowbar_act(mob/living/user, obj/item/I)
+	return
```
Comment (Rohesie) [inline code/game/atoms.dm]:
```suggestion
	return FALSE
```
For clarity, all of these should return a boolean.

File: code/game/objects/items/tools/maintenance_tools.dm (hunk 1 above)
Comment (Jalleo) [reply code/game/objects/items/tools/maintenance_tools.dm]: Whoops fixing now. I will leave a commented message at the top of this file saying that toolspeed affects how fast it works.

File: code/_onclick/item_attack.dm [hunk 7]
```diff
@@ -1,4 +1,21 @@
 
+/obj/item/proc/melee_attack_chain(mob/user, atom/target, params)
+	if(!tool_attack_chain(user, target) && pre_attack(target, user, params))
+		// Return 1 in attackby() to prevent afterattack() effects (when safely moving items for example)
+		var/resolved = target.attackby(src, user, params)
+		if(!resolved && target && !QDELETED(src))
+			afterattack(target, user, 1, params) // 1: clicking something Adjacent
```
Comment (Rohesie) [inline code/_onclick/item_attack.dm]:
```suggestion
			afterattack(target, user, TRUE, params) // 1: clicking something Adjacent
```

File: code/_onclick/item_attack.dm [hunk 8]
```diff
@@ -1,4 +1,21 @@
 
+/obj/item/proc/melee_attack_chain(mob/user, atom/target, params)
+	if(!tool_attack_chain(user, target) && pre_attack(target, user, params))
```
Comment (Rohesie) [inline code/_onclick/item_attack.dm]: Do an early return here.

File: code/_onclick/item_attack.dm [hunk 9]
```diff
@@ -7,6 +24,10 @@
 
 
 
+/obj/item/proc/pre_attack(atom/A, mob/living/user, params) //do stuff before attackby!
```
Comment (Rohesie) [inline code/_onclick/item_attack.dm]: This proc is not really necessary. Remove it, same as the check, until it's actually needed/useful.

File: code/_onclick/item_attack.dm (hunk 8 above)
Comment (Jalleo) [reply code/_onclick/item_attack.dm]: I cant this is required to be processed to go to afterattack() now

File: code/_onclick/item_attack.dm (hunk 9 above)
Comment (Jalleo) [reply code/_onclick/item_attack.dm]: Shall I comment it out for now then?

Comment (Rohesie) [reply code/_onclick/item_attack.dm]: Remove instead.

File: code/_onclick/item_attack.dm (hunk 8 above)
Comment (Rohesie) [reply code/_onclick/item_attack.dm]: Explained what an early return without altering functionality is through discord.

File: code/_onclick/item_attack.dm [hunk 10]
```diff
@@ -1,12 +1,28 @@
 
+/obj/item/proc/melee_attack_chain(mob/user, atom/target, params)
+	if(tool_attack_chain(user, target))
+		return
+	// Return 1 in attackby() to prevent afterattack() effects (when safely moving items for example)
+	var/resolved = target.attackby(src, user, params)
+	if(!resolved && target && !QDELETED(src))
```
Comment (Rohesie) [inline code/_onclick/item_attack.dm]:
```suggestion
	if(resolved || QDELETED(target) || QDELETED(src))
		return
	afterattack(target, user, TRUE, params) // TRUE: clicking something Adjacent
```

Comment (Jalleo) [reply code/_onclick/item_attack.dm]: are you sure about the QDELETED(target) ?

Comment (Rohesie) [reply code/_onclick/item_attack.dm]: Yep, sure!

File: code/game/objects/items.dm [hunk 11]
```diff
@@ -821,3 +827,73 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
...
+
+	// Play tool sound at the beginning of tool usage.
+	play_tool_sound(target, volume)
+
+	if(delay)
+		// Create a callback with checks that would be called every tick by do_after.
+		var/datum/callback/tool_check = CALLBACK(src, .proc/tool_check_callback, user, amount, extra_checks)
+
+		if(ismob(target))
+			if(!do_mob(user, target, delay, extra_checks=tool_check))
```
Comment (DominikPanic) [inline code/game/objects/items.dm]: Combine these two

File: code/game/objects/items.dm [hunk 12]
```diff
@@ -821,3 +827,73 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
...
+				return
+
+		else
+			if(!do_after(user, delay, target=target, extra_checks=tool_check))
```
Comment (DominikPanic) [inline code/game/objects/items.dm]: Also combine, here into else if

File: code/game/objects/items.dm [hunk 13]
```diff
@@ -821,3 +827,73 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
...
+				return
+	else
+		// Invoke the extra checks once, just in case.
+		if(extra_checks && !extra_checks.Invoke())
```
Comment (DominikPanic) [inline code/game/objects/items.dm]: Also else if

File: code/game/objects/items.dm [hunk 14]
```diff
@@ -821,3 +827,73 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
...
+	return !amount
+
+// Generic use proc. Depending on the item, it uses up fuel, charges, sheets, etc.
+// Returns TRUE on success, FALSE on failure.
+/obj/item/proc/use(used)
+	return !used
+
+// Plays item's usesound, if any.
+/obj/item/proc/play_tool_sound(atom/target, volume)
+	if(target && usesound && volume)
```
Comment (DominikPanic) [inline code/game/objects/items.dm]: Early return?

**Comment by Jalleo** [comment]
Okay I will fix these bits tomorrow evening. Taking a little break.

File: code/game/objects/items.dm [hunk 15]
```diff
@@ -846,17 +846,13 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
...
+		if(ismob(target) && !do_mob(user, target, delay, extra_checks=tool_check))
+			return
 
-		else
-			if(!do_after(user, delay, target=target, extra_checks=tool_check))
-				return
-	else
-		// Invoke the extra checks once, just in case.
-		if(extra_checks && !extra_checks.Invoke())
+		else if(!do_after(user, delay, target=target, extra_checks=tool_check))
```
Comment (Rohesie) [inline code/game/objects/items.dm]: This logic is problematic. This `else` will trigger for both when the target is not a mob (as should be) but also if it fails the `do_mob` (that is not intended).
Keep the old indentation on the `ismob` and `do_mob` checks.

Comment (Jalleo) [reply code/game/objects/items.dm]: I have fixed those issues now

**Comment by Rohesie** [comment]
Travis is acting up. Is this working fine locally?

**Comment by Jalleo** [comment]
Travis has built fine?

**Pull request merged**